[flang][OpenMP] Compilation error of the same COMMON block specified in two THREADPRIVATE directive clauses

```
Version of flang-new : 17.0.0(0303eafcb34647f7d5a4015aad266b5766f5dc5e)
```

The same `COMMON` block specified in two `THREADPRIVATE` directives is a compile-time error.

The following are the test program, flang-new, gfortran and ifort compilation result.

omp30_example_a_25_04_.f90:
```fortran
program main
  call A25_4_GOOD()
  print *,'pass'
end program main

SUBROUTINE A25_4_GOOD()
  COMMON /T/ A
!$OMP THREADPRIVATE(/T/)
  integer :: A
  A=100
  call A25_4S_GOOD()
  if (A.ne.101) print *,'err'
CONTAINS
  SUBROUTINE A25_4S_GOOD()
    COMMON /T/ A
!$OMP THREADPRIVATE(/T/)
    integer :: A
!$OMP PARALLEL COPYIN(/T/)
    A=A+1
!$OMP END PARALLEL
  END SUBROUTINE A25_4S_GOOD
END SUBROUTINE A25_4_GOOD
```

```
$ flang-new -flang-experimental-exec -fopenmp omp30_example_a_25_04_.f90
error: Semantic errors in omp30_example_a_25_04_.f90
./omp30_example_a_25_04_.f90:16:22: error: 't' appears in more than one data-sharing clause on the same OpenMP directive
  !$OMP THREADPRIVATE(/T/)
                       ^
./omp30_example_a_25_04_.f90:18:24: error: Non-THREADPRIVATE object 'a' in COPYIN clause
  !$OMP PARALLEL COPYIN(/T/)
                         ^
$
```

```
$ gfortran -fopenmp omp30_example_a_25_04_.f90
$
```

```
$ ifort -qopenmp omp30_example_a_25_04_.f90
$
```

--- a/flang/lib/Semantics/resolve-directives.cpp
+++ b/flang/lib/Semantics/resolve-directives.cpp
@@ -1944,17 +1944,18 @@ void OmpAttributeVisitor::ResolveOmpNameList(
 
 Symbol *OmpAttributeVisitor::ResolveOmpCommonBlockName(
     const parser::Name *name) {
-  if (auto *prev{name
-              ? GetContext().scope.parent().FindCommonBlock(name->source)
-              : nullptr}) {
+  if (!name) {
+    return nullptr;
+  }
+  // First check if the Common Block is declared in the current scope
+  if (auto *cur{GetContext().scope.FindCommonBlock(name->source)}) {
+    name->symbol = cur;
+    return cur;
+  }
+  // Then check parent scope
+  if (auto *prev{GetContext().scope.parent().FindCommonBlock(name->source)}) {
     name->symbol = prev;
     return prev;
-  }
-  // Check if the Common Block is declared in the current scope
-  if (auto *commonBlockSymbol{
-          name ? GetContext().scope.FindCommonBlock(name->source) : nullptr}) {
-    name->symbol = commonBlockSymbol;
-    return commonBlockSymbol;
   }
   return nullptr;
 }

--- a/flang/test/Semantics/OpenMP/threadprivate06.f90
+++ b/flang/test/Semantics/OpenMP/threadprivate06.f90
@@ -1,0 +1,30 @@
+! RUN: %python %S/../test_errors.py %s %flang_fc1 -fopenmp
+! OpenMP Version 5.1
+! Check OpenMP construct validity for the following directives:
+! 2.21.2 Threadprivate Directive
+
+program main
+  call sub1()
+  print *, 'pass'
+end program main
+
+subroutine sub1()
+  common /c/ a
+  !$omp threadprivate(/c/)
+  integer :: a
+
+  a = 100
+  call sub2()
+  if (a .ne. 101) print *, 'err'
+
+contains
+  subroutine sub2()
+    common /c/ a
+    !$omp threadprivate(/c/)
+    integer :: a
+
+    !$omp parallel copyin(/c/)
+      a = a + 1
+    !$omp end parallel
+  end subroutine
+end subroutine

